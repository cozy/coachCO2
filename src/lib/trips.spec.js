import {
  makeBicycleTrip,
  makeWalkingTrip,
  mockFeature,
  mockFeatureCollection,
  mockSerie,
  makeCarTrip,
  modeProps
} from 'test/mockTrip'

import {
  computeAndformatCaloriesTrip,
  getSectionsInfo,
  getPurpose,
  getSectionsFormatedInfo,
  getModesSortedByDistance,
  computeAndFormatCO2Trip
} from 'src/lib/trips'

const mockedFeatures = () => [
  mockFeature('featureId01'),
  mockFeature('featureId02'),
  mockFeature('featureId03'),
  mockFeature('featureId04'),
  mockFeatureCollection('featureCollectionId01', [
    mockFeature('featureId05', modeProps.bicycle)
  ]),
  mockFeatureCollection('featureCollectionId02', [
    mockFeature('featureId06', modeProps.walking)
  ]),
  mockFeatureCollection('featureCollectionId03', [
    mockFeature('featureId07', modeProps.car)
  ])
]

describe('getModesSortedByDistance', () => {
  const mockedSerie = mockSerie('serieId01', mockedFeatures())

  it('should return feature collection modes sorted by distance', () => {
    const result = getModesSortedByDistance(mockedSerie)

    expect(result).toHaveLength(3)
    expect(result).toStrictEqual(['CAR', 'BICYCLING', 'WALKING'])
  })
})

describe('computeAndformatCaloriesTrip', () => {
  it('should return formated value', () => {
    const bCalories = computeAndformatCaloriesTrip(makeBicycleTrip())
    expect(bCalories).toBe('104 kcal')

    const wCalories = computeAndformatCaloriesTrip(makeWalkingTrip())
    expect(wCalories).toBe('23 kcal')
  })
})

describe('computeAndFormatCO2Trip', () => {
  it('should return formated value', () => {
    const bCO2 = computeAndFormatCO2Trip(makeBicycleTrip())
    expect(bCO2).toBe('0 kg')

    const cCO2 = computeAndFormatCO2Trip(makeCarTrip())
    expect(cCO2).toBe('2.84 kg')
  })
})

describe('getPurpose', () => {
  it('should return purpose value', () => {
    const trips = mockSerie('serieId01', mockedFeatures(), {
      manual_purpose: 'PICK_DROP'
    })
    const purpose = getPurpose(trips)

    expect(purpose).toBe('PICK_DROP')
  })
})

describe('getSectionsInfo', () => {
  const trips = mockSerie('serieId01', mockedFeatures())

  it.each`
    trips    | property
    ${trips} | ${'averageSpeed'}
    ${trips} | ${'coordinates'}
    ${trips} | ${'distance'}
    ${trips} | ${'distances'}
    ${trips} | ${'duration'}
    ${trips} | ${'endDate'}
    ${trips} | ${'id'}
    ${trips} | ${'mode'}
    ${trips} | ${'speeds'}
    ${trips} | ${'startDate'}
    ${trips} | ${'timestamps'}
  `(`should section must have $property property`, ({ trips, property }) => {
    const sectionsInfo = getSectionsInfo(trips)
    expect(sectionsInfo[0]).toHaveProperty(property)
  })

  it('should return correct sections', () => {
    const sectionsInfo = getSectionsInfo(trips)
    expect(sectionsInfo).toMatchSnapshot()
  })
})

describe('getSectionsFormatedInfo', () => {
  it('should return formated value', () => {
    const bicyleInfos = getSectionsFormatedInfo(makeBicycleTrip(), 'en')
    expect(bicyleInfos[0]).toMatchObject({
      distance: '2 km',
      duration: '10 min',
      averageSpeed: '16 km/h'
    })
  })
})
