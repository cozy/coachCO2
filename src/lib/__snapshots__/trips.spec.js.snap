// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSectionsInfo should return correct sections 1`] = `
Array [
  Object {
    "averageSpeed": 16.158880604162004,
    "coordinates": Array [
      Array [
        2.31251,
        48.7799432,
      ],
      Array [
        2.313591181324194,
        48.78161416932299,
      ],
      Array [
        2.314672362648388,
        48.783285138645994,
      ],
    ],
    "distance": 2456,
    "distances": Array [
      0,
      201.98652472578271,
      201.98549037826737,
    ],
    "duration": 600,
    "endDate": 2021-12-07T15:39:16.000Z,
    "id": "featureId05",
    "mode": "BICYCLING",
    "speeds": Array [
      0,
      6.73288415752609,
      6.732849679275579,
    ],
    "startDate": 2021-12-07T15:24:49.000Z,
    "timestamps": Array [
      1638890689000,
      1638890719000,
      1638890749000,
    ],
  },
  Object {
    "averageSpeed": 0.3289624723613862,
    "coordinates": Array [
      Array [
        2.31251,
        48.7799432,
      ],
      Array [
        2.313591181324194,
        48.78161416932299,
      ],
      Array [
        2.314672362648388,
        48.783285138645994,
      ],
    ],
    "distance": 563,
    "distances": Array [
      0,
      2.3338746642432437,
      5.89018714479141,
    ],
    "duration": 540,
    "endDate": 2021-12-07T16:04:07.000Z,
    "id": "featureId06",
    "mode": "WALKING",
    "speeds": Array [
      0,
      0.07779582214144146,
      0.196339571493047,
    ],
    "startDate": 2021-12-07T15:39:31.000Z,
    "timestamps": Array [
      1638891571000,
      1638891601000,
      1638891631000,
    ],
  },
  Object {
    "averageSpeed": 2.9874387058354404,
    "coordinates": Array [
      Array [
        2.31251,
        48.7799432,
      ],
      Array [
        2.313591181324194,
        48.78161416932299,
      ],
      Array [
        2.314672362648388,
        48.783285138645994,
      ],
    ],
    "distance": 14789,
    "distances": Array [
      0,
      43.094121726730805,
      31.5918459191552,
    ],
    "duration": 1800,
    "endDate": 2021-12-07T16:06:26.000Z,
    "id": "featureId07",
    "mode": "CAR",
    "speeds": Array [
      0,
      1.4364707242243602,
      1.0530615306385067,
    ],
    "startDate": 2021-12-07T16:04:34.000Z,
    "timestamps": Array [
      1638893074000,
      1638893104000,
      1638893134000,
    ],
  },
]
`;
